code/GatewayFieldsFactory.php

```diff
@@ -247,69 +243,93 @@ public function getCompanyFields()
     }
 
     /**
+     * Clear all fields that are not required by the gateway. Does nothing if gateway is null
+     * @param $fields
+     * @param array $defaults
+     */
+    protected function cullForGateway(&$fields, $defaults = array())
+    {
+        if (!$this->gateway){
+            return;
+        }
+
+        $selected = array_merge($defaults, GatewayInfo::requiredFields($this->gateway));
+        foreach ($fields as $name => $field) {
+            if (!in_array($name, $selected)) {
+                unset($fields[$name]);
+            }
+        }
+    }
+
+    /**
      * Attempts to find a custom field name and/or prefix defined in rename.yml, otherwise returns the same input
      * that it was given
      *
-     * @param string|array  $defaultName The default name of the field
-     * @param null|bool     $skipGateway Used in recursion
+     * @param string|array $defaultName The default name of the field
+     * @param null|bool    $skipGateway Used in recursion
      *
      * @return string|array
      */
-    public function getFieldName($defaultName, $skipGateway = null) {
-        // batch support
+    public function getFieldName($defaultName, $skipGateway = null)
```

Comment:
As explained below, this method does too much. The rename-map should be precalculated and then this method should just look up the value from the rename-map.